includes/SVG.php

```diff
@@ -0,0 +1,428 @@
+			$filetypes[] = self::EXT;
+			$value       = implode( ' ', $filetypes );
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Hook into metadata generation and get height and width for SVG file.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param array  $metadata      An array of attachment meta data.
+	 * @param int    $attachment_id Current attachment ID.
+	 * @param string $context       Additional context. Can be 'create' when metadata was initially created for new
+	 *                              attachment.
+	 *
+	 * @return array
+	 */
+	public function wp_generate_attachment_metadata( $metadata, $attachment_id, $context ) {
+		if ( 'create' !== $context ) {
+			return $metadata;
+		}
+		$attachment = get_post( $attachment_id );
+		$mime_type  = get_post_mime_type( $attachment );
+
+		if ( self::MIME_TYPE !== $mime_type ) {
+			return $metadata;
+		}
+		$file = get_attached_file( $attachment_id );
+		if ( false === $file ) {
+			return $metadata;
+		}
+
+		$size = $this->get_svg_size( $file );
+		// Check if image size failed to generate and return if so.
+		if ( is_wp_error( $size ) ) {
+			return $metadata;
+		}
+
+		return [
+			'width'    => $size['width'],
+			'height'   => $size['height'],
+			'file'     => _wp_relative_upload_path( $file ),
+			'filesize' => (int) filesize( $file ),
+			'sizes'    => [],
+		];
+	}
+
+	/**
+	 * Hook into upload and error if size could not be generated.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @param array $upload {
+	 *      Array of upload data.
+	 *
+	 *      @type string $file   Filename of the newly-uploaded file.
+	 *      @type string $url    URL of the newly-uploaded file.
+	 *      @type string $type   Mime type of the newly-uploaded file.
+	 * }
+	 *
+	 * @return string[]
+	 */
+	public function wp_handle_upload( $upload ) {
+		if ( self::MIME_TYPE !== $upload['type'] ) {
+			return $upload;
+		}
+
+		$sanitized = $this->sanitize( $upload['tmp_name'] );
+		if ( is_wp_error( $sanitized ) ) {
+			return [ 'error' => $sanitized->get_error_message() ];
+		}
+
+		$size = $this->get_svg_size( $upload['tmp_name'] );
+		if ( is_wp_error( $size ) ) {
+			return [ 'error' => $size->get_error_message() ];
+		}
+
+		return $upload;
+	}
+
+
+	/**
+	 * Get SVG image size.
+	 *
+	 * @since 1.2.0
+	 *
+	 * @SuppressWarnings(PHPMD.NPathComplexity)
+	 *
+	 * @param string $file Path to SVG file.
+	 *
+	 * @return array|WP_Error
+	 */
+	protected function get_svg_size( $file ) {
+		$svg = $this->get_svg_data( $file );
+		$xml = $this->get_xml( $svg );
+
+		if ( false === $xml ) {
+			return new WP_Error( 'invalid_xml_svg', __( 'Invalid xml in SVG.', 'web-stories' ) );
+		}
+
+		$width  = (int) $xml->getAttribute( 'width' );
+		$height = (int) $xml->getAttribute( 'height' );
+
+		// If height and width are not set, try the viewport attribute.
+		if ( ! $width || ! $height ) {
+			$view_box = $xml->getAttribute( 'viewBox' );
+			if ( empty( $view_box ) ) {
+				$view_box = $xml->getAttribute( 'viewbox' );
```

Comment:
I thought only `viewBox` is valid? Should we even check for `viewbox`?

Reply:
In my testing, I found many svg in the wild using this lower case. I think whatever generated these files made invalud svg. But either way we kind of hard to support it.